File: assets/codebase/menu/PlatformInfoTextComponent.cs
namespace city.menu {
    /// <summary>
    /// Applies the current platform name and version to the demo-disc menu overlay text.
    /// </summary>
    public sealed class PlatformInfoTextComponent : UpdateComponent {
        /// <summary>
        /// Owning entity that hosts the platform-info overlay hierarchy.
        /// </summary>
        Entity OwnerEntity;

        /// <summary>
        /// Cached child entity that renders the platform name.
        /// </summary>
        Entity PlatformNameTextEntity;
        /// <summary>
        /// Cached child entity that renders the platform version.
        /// </summary>
        Entity PlatformVersionTextEntity;
        /// <summary>
        /// Cached text component that renders the platform name.
        /// </summary>
        TextComponent PlatformNameTextComponent;
        /// <summary>
        /// Cached layout component that anchors the platform name text.
        /// </summary>
        LayoutComponent PlatformNameLayoutComponent;
        /// <summary>
        /// Cached text component that renders the platform version.
        /// </summary>
        TextComponent PlatformVersionTextComponent;
        /// <summary>
        /// Cached layout component that anchors the platform version text.
        /// </summary>
        LayoutComponent PlatformVersionLayoutComponent;

        /// <summary>
        /// Captures the authored local position assigned to the platform-name text entity before runtime layout adjusts it.
        /// </summary>
        float3 PlatformNameBaseLocalPosition;

        /// <summary>
        /// Captures the authored layout size assigned to the platform-name text entity before runtime text measurement adjusts it.
        /// </summary>
        int2 PlatformNameBaseSize;

        /// <summary>
        /// Captures the authored local position assigned to the platform-version text entity before runtime layout adjusts it.
        /// </summary>
        float3 PlatformVersionBaseLocalPosition;

        /// <summary>
        /// Captures the authored layout size assigned to the platform-version text entity before runtime text measurement adjusts it.
        /// </summary>
        int2 PlatformVersionBaseSize;

        /// <summary>
        /// Tracks whether the runtime overlay hierarchy has been bound successfully.
        /// </summary>
        bool IsInitialized;

        /// <summary>
        /// Captures the owning entity and attempts initialization once the runtime hierarchy becomes available.
        /// </summary>
        /// <param name="entity">Owning initialized entity.</param>
        public override void ComponentAdded(Entity entity) {
            base.ComponentAdded(entity);
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            }

            OwnerEntity = entity;
            IsInitialized = false;
            ClearBinding();
            TryInitialize();
        }

        /// <summary>
        /// Releases cached runtime-only binding state when the component leaves the scene hierarchy.
        /// </summary>
        /// <param name="entity">Owning entity that is removing this component.</param>
        public override void ComponentRemoved(Entity entity) {
            ClearBinding();
            OwnerEntity = null;
            IsInitialized = false;
            base.ComponentRemoved(entity);
        }

        /// <summary>
        /// Waits until the runtime scene loader has attached the child text hierarchy, then applies the live platform information.
        /// </summary>
        public override void Update() {
            if (!IsInitialized) {
                TryInitialize();
            }
        }

        /// <summary>
        /// Attempts to bind the overlay subtree once the scene loader has attached all child entities.
        /// </summary>
        void TryInitialize() {
            if (OwnerEntity == null) {
                return;
            }
            if (!TryBindTextEntities(OwnerEntity)) {
                return;
            } else if (!AreTextComponentsReadyForLayout()) {
                return;
            }

            ApplyCurrentPlatformInfo();
            IsInitialized = true;
        }

        /// <summary>
        /// Returns whether the bound overlay text components have the runtime font assets required for measurement and layout.
        /// </summary>
        /// <returns><c>true</c> when both cached text components expose non-null runtime fonts; otherwise <c>false</c>.</returns>
        bool AreTextComponentsReadyForLayout() {
            if (PlatformNameTextComponent == null || PlatformVersionTextComponent == null) {
                return false;
            }

            return PlatformNameTextComponent.Font != null
                && PlatformVersionTextComponent.Font != null;
        }

        /// <summary>
        /// Applies the current runtime platform name and version to the two overlay rows.
        /// </summary>
        void ApplyCurrentPlatformInfo() {
            if (Core.Instance == null) {
                throw new InvalidOperationException("Platform info requires an active Core instance.");
            } else if (Core.Instance.PlatformInfo == null) {
                throw new InvalidOperationException("Platform info requires initialized runtime platform metadata.");
            }

            bool useHorizontalRowLayout = Math.Abs(PlatformNameBaseLocalPosition.Y - PlatformVersionBaseLocalPosition.Y) < 0.01f;
            if (useHorizontalRowLayout) {
                ApplyHorizontalText(PlatformNameTextEntity, PlatformNameTextComponent, PlatformNameLayoutComponent, Core.Instance.PlatformInfo.Name, PlatformNameBaseLocalPosition.X, PlatformNameBaseLocalPosition.Y, PlatformNameBaseSize, TextAlignment.Left);
                ApplyHorizontalText(PlatformVersionTextEntity, PlatformVersionTextComponent, PlatformVersionLayoutComponent, Core.Instance.PlatformInfo.Version, PlatformVersionBaseLocalPosition.X, PlatformVersionBaseLocalPosition.Y, PlatformVersionBaseSize, TextAlignment.Right);
                return;
            }

            ApplyText(PlatformNameTextEntity, PlatformNameTextComponent, Core.Instance.PlatformInfo.Name, 0f);
            ApplyText(PlatformVersionTextEntity, PlatformVersionTextComponent, Core.Instance.PlatformInfo.Version, PlatformNameTextComponent.Size.Y + 6f);
        }

        /// <summary>
        /// Resolves and caches the first two descendant text entities beneath the overlay host.
        /// </summary>
        /// <param name="entity">Owning initialized entity.</param>
        /// <returns><c>true</c> when the platform-info rows were bound successfully; otherwise <c>false</c>.</returns>
        bool TryBindTextEntities(Entity entity) {
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            }

            Entity firstTextEntity = null;
            Entity secondTextEntity = null;
            if (!TryCollectFirstTwoTextEntities(entity, ref firstTextEntity, ref secondTextEntity)) {
                return false;
            }

            PlatformNameTextEntity = firstTextEntity;
            PlatformVersionTextEntity = secondTextEntity;
            PlatformNameTextComponent = FindTextComponent(PlatformNameTextEntity);
            PlatformVersionTextComponent = FindTextComponent(PlatformVersionTextEntity);
            PlatformNameLayoutComponent = FindLayoutComponent(PlatformNameTextEntity);
            PlatformVersionLayoutComponent = FindLayoutComponent(PlatformVersionTextEntity);
            PlatformNameBaseLocalPosition = PlatformNameTextEntity.LocalPosition;
            PlatformNameBaseSize = PlatformNameTextComponent.Size;
            PlatformVersionBaseLocalPosition = PlatformVersionTextEntity.LocalPosition;
            PlatformVersionBaseSize = PlatformVersionTextComponent.Size;
            return true;
        }

        /// <summary>
        /// Collects the first two descendant entities that carry a text component in stable depth-first order without allocating a temporary runtime list.
        /// </summary>
        /// <param name="parentEntity">Current subtree root to scan.</param>
        /// <param name="firstTextEntity">First discovered descendant text entity.</param>
        /// <param name="secondTextEntity">Second discovered descendant text entity.</param>
        /// <returns><c>true</c> when two descendant text entities were discovered; otherwise <c>false</c>.</returns>
        bool TryCollectFirstTwoTextEntities(Entity parentEntity, ref Entity firstTextEntity, ref Entity secondTextEntity) {
            if (parentEntity == null) {
                throw new ArgumentNullException(nameof(parentEntity));
            } else if (parentEntity.Children == null) {
                return false;
            }

            for (int childIndex = 0; childIndex < parentEntity.Children.Count; childIndex++) {
                Entity childEntity = parentEntity.Children[childIndex];
                if (childEntity == null) {
                    continue;
                }

                if (TryFindTextComponent(childEntity, out TextComponent textComponent)) {
                    if (firstTextEntity == null) {
                        firstTextEntity = childEntity;
                    } else if (secondTextEntity == null) {
                        secondTextEntity = childEntity;
                        return true;
                    }
                }

                if (TryCollectFirstTwoTextEntities(childEntity, ref firstTextEntity, ref secondTextEntity)) {
                    return true;
                }
            }

            return false;
        }

        /// <summary>
        /// Clears every cached runtime overlay reference.
        /// </summary>
        void ClearBinding() {
            PlatformNameTextEntity = null;
            PlatformVersionTextEntity = null;
            PlatformNameTextComponent = null;
            PlatformVersionTextComponent = null;
            PlatformNameLayoutComponent = null;
            PlatformVersionLayoutComponent = null;
            PlatformNameBaseLocalPosition = float3.Zero;
            PlatformNameBaseSize = new int2(0, 0);
            PlatformVersionBaseLocalPosition = float3.Zero;
            PlatformVersionBaseSize = new int2(0, 0);
        }

        /// <summary>
        /// Applies one platform-info text line to one child entity.
        /// </summary>
        /// <param name="entity">Child text entity that should be updated.</param>
        /// <param name="textComponent">Text component that renders the value.</param>
        /// <param name="text">Text content to display.</param>
        /// <param name="topOffset">Vertical offset from the overlay container.</param>
        void ApplyText(Entity entity, TextComponent textComponent, string text, float topOffset) {
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            } else if (textComponent == null) {
                throw new ArgumentNullException(nameof(textComponent));
            }

            textComponent.Alignment = TextAlignment.Right;
            textComponent.Text = text;
            float2 measuredSize = textComponent.Font.MeasureString(text);
            double fontScale = textComponent.FontScale;
            textComponent.Size = new int2(
                (int)Math.Ceiling(measuredSize.X * fontScale),
                (int)Math.Ceiling(measuredSize.Y * fontScale));
            entity.LocalPosition = new float3(-textComponent.Size.X, topOffset, 0f);
        }

        /// <summary>
        /// Applies one platform-info text line using authored horizontal anchors so the overlay can present one left/right row.
        /// </summary>
        /// <param name="entity">Child text entity that should be updated.</param>
        /// <param name="textComponent">Text component that renders the value.</param>
        /// <param name="text">Text content to display.</param>
        /// <param name="baseX">Authored local X anchor.</param>
        /// <param name="baseY">Authored local Y anchor.</param>
        /// <param name="baseSize">Authored layout box preserved for anchored horizontal alignment.</param>
        /// <param name="alignment">Requested horizontal text alignment.</param>
        void ApplyHorizontalText(Entity entity, TextComponent textComponent, LayoutComponent layoutComponent, string text, float baseX, float baseY, int2 baseSize, TextAlignment alignment) {
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            } else if (textComponent == null) {
                throw new ArgumentNullException(nameof(textComponent));
            }

            textComponent.Alignment = alignment;
            textComponent.Text = text;
            float2 measuredSize = textComponent.Font.MeasureString(text);
            double fontScale = textComponent.FontScale;
            textComponent.Size = new int2(
                Math.Max(baseSize.X, (int)Math.Ceiling(measuredSize.X * fontScale)),
                Math.Max(baseSize.Y, (int)Math.Ceiling(measuredSize.Y * fontScale)));
            if (layoutComponent != null && layoutComponent.IsAnchored) {
                layoutComponent.RefreshAnchoring();
                return;
            }
            if (alignment == TextAlignment.Right) {
                entity.LocalPosition = new float3(baseX - textComponent.Size.X, baseY, 0f);
                return;
            }

            entity.LocalPosition = new float3(baseX, baseY, 0f);
        }

        /// <summary>
        /// Finds the layout component attached to one child entity.
        /// </summary>
        /// <param name="entity">Child entity whose layout component should be searched.</param>
        /// <returns>Resolved layout component.</returns>
        LayoutComponent FindLayoutComponent(Entity entity) {
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            }

            for (int index = 0; index < entity.Components.Count; index++) {
                if (entity.Components[index] is LayoutComponent layoutComponent) {
                    return layoutComponent;
                }
            }

            return null;
        }

        /// <summary>
        /// Finds the text component attached to one child entity.
        /// </summary>
        /// <param name="entity">Child entity whose text component should be returned.</param>
        /// <returns>Attached text component.</returns>
        TextComponent FindTextComponent(Entity entity) {
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            }

            for (int index = 0; index < entity.Components.Count; index++) {
                if (entity.Components[index] is TextComponent textComponent) {
                    return textComponent;
                }
            }

            throw new InvalidOperationException("Platform-info overlay child must include a text component.");
        }

        /// <summary>
        /// Attempts to find a text component attached to one child entity.
        /// </summary>
        /// <param name="entity">Child entity whose text component should be searched.</param>
        /// <param name="textComponent">Resolved text component when one exists.</param>
        /// <returns><c>true</c> when the entity contains a text component; otherwise <c>false</c>.</returns>
        bool TryFindTextComponent(Entity entity, out TextComponent textComponent) {
            if (entity == null) {
                throw new ArgumentNullException(nameof(entity));
            }

            for (int index = 0; index < entity.Components.Count; index++) {
                if (entity.Components[index] is TextComponent foundTextComponent) {
                    textComponent = foundTextComponent;
                    return true;
                }
            }

            textComponent = null;
            return false;
        }
    }
}

File: windows-build-20260602-directional-shadow-plaza-direct/cooked/code/gameplay/PlatformInfoTextComponent.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "PlatformInfoTextComponent.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"
#include "runtime/native_string.hpp"
#include "PlatformInfoTextComponent.hpp"
#include "system/math.hpp"
#include "Component.hpp"
#include "Core.hpp"
#include "Entity.hpp"
#include "FontAsset.hpp"
#include "PlatformInfo.hpp"
#include "TextAlignment.hpp"
#include "TextComponent.hpp"
#include "UpdateComponent.hpp"
#include "float2.hpp"
#include "float3.hpp"
#include "int2.hpp"
#include "system/math.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"

PlatformInfoTextComponent::PlatformInfoTextComponent() : IsInitialized(), OwnerEntity(), PlatformNameBaseLocalPosition(), PlatformNameTextComponent(), PlatformNameTextEntity(), PlatformVersionBaseLocalPosition(), PlatformVersionTextComponent(), PlatformVersionTextEntity()
{
}

void PlatformInfoTextComponent::ComponentAdded(Entity* entity)
{
UpdateComponent::ComponentAdded(entity);
    if (entity == nullptr)
    {
throw new ArgumentNullException("entity");
    }
this->OwnerEntity = entity;
this->IsInitialized = false;
this->ClearBinding();
this->TryInitialize();
}

void PlatformInfoTextComponent::ComponentRemoved(Entity* entity)
{
this->ClearBinding();
this->OwnerEntity = nullptr;
this->IsInitialized = false;
UpdateComponent::ComponentRemoved(entity);
}

void PlatformInfoTextComponent::Update()
{
    if (!this->IsInitialized)
    {
this->TryInitialize();
    }
}

void PlatformInfoTextComponent::ApplyCurrentPlatformInfo()
{
    if (Core::get_Instance() == nullptr)
    {
throw new InvalidOperationException("Platform info requires an active Core instance.");
    }
else {
    if (Core::get_Instance()->get_PlatformInfo() == nullptr)
    {
throw new InvalidOperationException("Platform info requires initialized runtime platform metadata.");
    }
}
const bool useHorizontalRowLayout = Math::Abs(this->PlatformNameBaseLocalPosition.Y - this->PlatformVersionBaseLocalPosition.Y) < 0.01f;
    if (useHorizontalRowLayout)
    {
this->ApplyHorizontalText(this->PlatformNameTextEntity, this->PlatformNameTextComponent, Core::get_Instance()->get_PlatformInfo()->get_Name(), this->PlatformNameBaseLocalPosition.X, this->PlatformNameBaseLocalPosition.Y, TextAlignment::Left);
this->ApplyHorizontalText(this->PlatformVersionTextEntity, this->PlatformVersionTextComponent, Core::get_Instance()->get_PlatformInfo()->get_Version(), this->PlatformVersionBaseLocalPosition.X, this->PlatformVersionBaseLocalPosition.Y, TextAlignment::Right);
return;    }
this->ApplyText(this->PlatformNameTextEntity, this->PlatformNameTextComponent, Core::get_Instance()->get_PlatformInfo()->get_Name(), 0.0f);
this->ApplyText(this->PlatformVersionTextEntity, this->PlatformVersionTextComponent, Core::get_Instance()->get_PlatformInfo()->get_Version(), this->PlatformNameTextComponent->get_Size().Y + 6.0f);
}

void PlatformInfoTextComponent::ApplyHorizontalText(Entity* entity, TextComponent* textComponent, std::string text, float baseX, float baseY, TextAlignment alignment)
{
    if (entity == nullptr)
    {
throw new ArgumentNullException("entity");
    }
else {
    if (textComponent == nullptr)
    {
throw new ArgumentNullException("textComponent");
    }
}
textComponent->set_Alignment(alignment);
textComponent->set_Text(text);
float2 measuredSize = textComponent->get_Font()->MeasureString(text);
const double fontScale = textComponent->get_FontScale();
textComponent->set_Size(([&]() {
auto __ctor_arg_0000009E = static_cast<int32_t>(Math::Ceiling(measuredSize.X * fontScale));
auto __ctor_arg_0000009F = static_cast<int32_t>(Math::Ceiling(measuredSize.Y * fontScale));
return int2(__ctor_arg_0000009E, __ctor_arg_0000009F);
})());
    if (alignment == TextAlignment::Right)
    {
entity->set_LocalPosition(float3(baseX - textComponent->get_Size().X, baseY, 0.0f));
return;    }
entity->set_LocalPosition(float3(baseX, baseY, 0.0f));
}

void PlatformInfoTextComponent::ApplyText(Entity* entity, TextComponent* textComponent, std::string text, float topOffset)
{
    if (entity == nullptr)
    {
throw new ArgumentNullException("entity");
    }
else {
    if (textComponent == nullptr)
    {
throw new ArgumentNullException("textComponent");
    }
}
textComponent->set_Alignment(TextAlignment::Right);
textComponent->set_Text(text);
float2 measuredSize = textComponent->get_Font()->MeasureString(text);
const double fontScale = textComponent->get_FontScale();
textComponent->set_Size(([&]() {
auto __ctor_arg_000000A0 = static_cast<int32_t>(Math::Ceiling(measuredSize.X * fontScale));
auto __ctor_arg_000000A1 = static_cast<int32_t>(Math::Ceiling(measuredSize.Y * fontScale));
return int2(__ctor_arg_000000A0, __ctor_arg_000000A1);
})());
entity->set_LocalPosition(float3(-textComponent->get_Size().X, topOffset, 0.0f));
}

bool PlatformInfoTextComponent::AreTextComponentsReadyForLayout()
{
    if (this->PlatformNameTextComponent == nullptr || this->PlatformVersionTextComponent == nullptr)
    {
return false;    }
return this->PlatformNameTextComponent->get_Font() != nullptr && this->PlatformVersionTextComponent->get_Font() != nullptr;}

void PlatformInfoTextComponent::ClearBinding()
{
this->PlatformNameTextEntity = nullptr;
this->PlatformVersionTextEntity = nullptr;
this->PlatformNameTextComponent = nullptr;
this->PlatformVersionTextComponent = nullptr;
this->PlatformNameBaseLocalPosition = float3::get_Zero();
this->PlatformVersionBaseLocalPosition = float3::get_Zero();
}

void PlatformInfoTextComponent::CollectChildTextEntities(Entity* parentEntity, List<Entity*>* textEntities)
{
    if (parentEntity == nullptr)
    {
throw new ArgumentNullException("parentEntity");
    }
else {
    if (textEntities == nullptr)
    {
throw new ArgumentNullException("textEntities");
    }
else {
    if (parentEntity->get_Children() == nullptr)
    {
return;    }
}
}
for (int32_t childIndex = 0; childIndex < parentEntity->get_Children()->get_Count(); childIndex++) {
Entity *childEntity = (*parentEntity->get_Children()).get_Item(childIndex);
    if (childEntity == nullptr)
    {
continue;
    }
TextComponent* textComponent;
    if (this->TryFindTextComponent__out1(childEntity, textComponent))
    {
textEntities->Add(childEntity);
    if (textEntities->get_Count() >= 2)
    {
return;    }
    }
this->CollectChildTextEntities(childEntity, textEntities);
    if (textEntities->get_Count() >= 2)
    {
return;    }
}
}

TextComponent* PlatformInfoTextComponent::FindTextComponent(Entity* entity)
{
    if (entity == nullptr)
    {
throw new ArgumentNullException("entity");
    }
for (int32_t index = 0; index < entity->get_Components()->get_Count(); index++) {
    TextComponent* textComponent = he_cpp_try_cast<TextComponent>((*entity->get_Components()).get_Item(index));
    if (textComponent != nullptr)
    {
return textComponent;    }
}
throw new InvalidOperationException("Platform-info overlay child must include a text component.");
}

bool PlatformInfoTextComponent::TryBindTextEntities(Entity* entity)
{
    if (entity == nullptr)
    {
throw new ArgumentNullException("entity");
    }
List<Entity*> *textEntities = new List<Entity*>();
this->CollectChildTextEntities(entity, textEntities);
    if (textEntities->get_Count() < 2)
    {
return false;    }
this->PlatformNameTextEntity = (*textEntities).get_Item(0);
this->PlatformVersionTextEntity = (*textEntities).get_Item(1);
this->PlatformNameTextComponent = this->FindTextComponent(this->PlatformNameTextEntity);
this->PlatformVersionTextComponent = this->FindTextComponent(this->PlatformVersionTextEntity);
this->PlatformNameBaseLocalPosition = this->PlatformNameTextEntity->get_LocalPosition();
this->PlatformVersionBaseLocalPosition = this->PlatformVersionTextEntity->get_LocalPosition();
return true;}

bool PlatformInfoTextComponent::TryFindTextComponent__out1(Entity* entity, TextComponent*& textComponent)
{
    if (entity == nullptr)
    {
throw new ArgumentNullException("entity");
    }
for (int32_t index = 0; index < entity->get_Components()->get_Count(); index++) {
    TextComponent* foundTextComponent = he_cpp_try_cast<TextComponent>((*entity->get_Components()).get_Item(index));
    if (foundTextComponent != nullptr)
    {
textComponent = foundTextComponent;
return true;    }
}
textComponent = nullptr;
return false;}

void PlatformInfoTextComponent::TryInitialize()
{
    if (this->OwnerEntity == nullptr)
    {
return;    }
    if (!this->TryBindTextEntities(this->OwnerEntity))
    {
return;    }
else {
    if (!this->AreTextComponentsReadyForLayout())
    {
return;    }
}
this->ApplyCurrentPlatformInfo();
this->IsInitialized = true;
}

